ratis-server/src/main/java/org/apache/ratis/server/storage/FileChunkReader.java

```diff
@@ -47,11 +51,14 @@ public class FileChunkReader implements Closeable {
    *
    * @param info the information of the file.
    * @param relativePath the relative path of the file.
+   * @param chunkMaxSize maximum chunk size for each {@link #readFileChunk()} call
    * @throws IOException if it failed to open the file.
    */
-  public FileChunkReader(FileInfo info, Path relativePath) throws IOException {
+  public FileChunkReader(FileInfo info, Path relativePath, int chunkMaxSize) throws IOException {
     this.info = info;
     this.relativePath = relativePath;
+    this.chunkMaxSize = Math.min(chunkMaxSize, (int) info.getFileSize());
+    this.chunkBuffer = new byte[chunkMaxSize];
```

Comment:
Since `ByteString` is supposed to be immutable, `FileChunkProto.Builder` won't copy the `data` in its `build()` method.  If we only have a single buffer, we can only have a single outstanding `FileChunkProto` (and a single outstanding `InstallSnapshotRequestProto`).  Otherwise, if we read a second `FileChunkProto`, the `data` in first `FileChunkProto` will be changed silently.  (That's why it is named `UnsafeByteOperations`.)